### PR TITLE
Handle '+' sign in url when using 'back' parameter to redirect

### PR DESCRIPTION
**Pull request by atomiix**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Since PR #20006 we allow '+' sign in urls but it was converted to a space character when redirecting from the `back` url parameter. This PR fixes it.
| Type?         | bug fix
| Category?     | BO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | 
| How to test?  | 1. Install PrestaShop in a subfolder containing the '+' sign (e.g. `http://locahost/prestashop+dev/`).<br>2. Go to Customers > Customers<br>3. Select any customer and go to the customer detail page<br>4. Scroll down to the bottom of the page and in the address block click on the + icon to add a new address<br>5. Fill all the required fields and click on save<br>6. You should be redirected correctly

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/21690)

### CODE REVIEW

**Review by Progi1984 — DISMISSED** [review]

**Review by matthieu-rolland — DISMISSED** [review]

**Comment by khouloudbelguith** [comment]
Hi @atomiix,

I manage to reproduce a similar issue not fixed by this PR:

Steps to reproduce the issue:  `Shop URL: http://prestashop.local/+17+8x+/`

1. Install PrestaShop in a subfolder containing the '+' sign
2. Go to BO > Discounts page
3. Create a Voucher for a specific customer (for example "John Doe") 
image
4. Go to Customers > Customers
5.  Select the customer  "John Doe" and go to the customer detail page
6. Try to edit the Voucher => see the error
7. Try to delete the voucher => see the error

image

Thanks!

File: classes/controller/AdminController.php [hunk 1]
```diff
@@ -2521,7 +2521,7 @@ public function renderForm()
             $helper->tpl_vars = $this->getTemplateFormVars();
             $helper->show_cancel_button = (isset($this->show_form_cancel_button)) ? $this->show_form_cancel_button : ($this->display == 'add' || $this->display == 'edit');
 
-            $back = urldecode(Tools::getValue('back', ''));
+            $back = urldecode(str_replace('+', '%2B', Tools::getValue('back', '')));
```
Comment (PierreRambaud) [inline classes/controller/AdminController.php]: Should it be easier to use rawurlencode instead?

Comment (atomiix) [reply classes/controller/AdminController.php]: Definitely 🤦‍♂️

**Review by Progi1984 — APPROVED** [review]

File: src/PrestaShopBundle/EventListener/BackUrlRedirectResponseListener.php [hunk 2]
```diff
@@ -79,7 +79,7 @@ public function onKernelResponse(FilterResponseEvent $event)
         $backUrl = $this->backUrlProvider->getBackUrl($currentRequest);
 
         if ($backUrl && !$this->isRequestUrlEqualToResponseUrl($currentRequest, $originalResponse)) {
-            $backUrlResponse = $originalResponse->setTargetUrl(urldecode($backUrl));
+            $backUrlResponse = $originalResponse->setTargetUrl($backUrl);
```
Comment (PierreRambaud) [inline src/PrestaShopBundle/EventListener/BackUrlRedirectResponseListener.php:82]: Woot, the url was decoded twice :scream:

**Review by PierreRambaud — APPROVED** [review]

**Comment by khouloudbelguith** [comment]
Hi @atomiix,

I found an issue following these steps:

1. Install PrestaShop in a subfolder containing the '+' sign
2. Go to BO > Discounts page
3. Create a Voucher for a specific customer (for example "John Doe") 
image
4. Go to Customers > Customers
5.  Select the customer  "John Doe" and go to the customer detail page
7. Try to delete the voucher => see the error, the alert displayed is _`Update successful.`_ instead of _`Successful deletion.`_

image


Thanks!

**Comment by atomiix** [comment]
@khouloudbelguith I don't think this is related to this PR. Are you sure this message is not already this one when you follow the same steps without having a `+` in the PrestaShop path?

**Comment by khouloudbelguith** [comment]
Hi @atomiix,

In the develop branch, without this PR, we don't have any alert
https://drive.google.com/file/d/1c__6MpRJdeeSfaRBDOFQvRRVi2fFOIO5/view?usp=sharing
As reported here: https://github.com/PrestaShop/PrestaShop/issues/18841

Thanks!

**Comment by atomiix** [comment]
@khouloudbelguith I just tried again and with my PR I don't get any message either. Are you on a clean install on develop?

**Comment by khouloudbelguith** [comment]
Hi @atomiix,

Yes, I make a new install with this PR
I attached a screen record
https://drive.google.com/file/d/1XWEqnjkmWoRHuQF-oZrTfCJoL_vSd_jl/view?usp=sharing
here's my configuration
image

I'm using Chrome Version 85.0.4183.102 (Build officiel) (64 bits)

Thanks!

**Comment by atomiix** [comment]
@khouloudbelguith I think I know what happened 😅  When you first edit the cart rule, you get the message "Update successful" that **stays** when you then delete the cart rule. But, if you try to delete it first, you should not get any message.

**Comment by khouloudbelguith** [comment]
Hi @atomiix,

Yes, you are right, thank you :smile: 
It is ok :heavy_check_mark: 
https://drive.google.com/file/d/13k0tc9Y0QXqQ1-wqTDm5pcd-B8kC6vHf/view?usp=sharing

Thanks!

**Pull request merged**